pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,501 @@
+			// leave it empty. We want to list this resource in all
+			// namespaces if it's namespace scoped, so leave
+			// APIResource.Namespaced as false is all right.
+			apiResource := metav1.APIResource{Name: resource.Resource}
+			return client.ParameterCodec(dynamic.VersionedParameterEncoderWithV1Fallback).
+				Resource(&apiResource, v1.NamespaceAll).
+				List(&options)
+		},
+		WatchFunc: func(options v1.ListOptions) (watch.Interface, error) {
+			// APIResource.Kind is not used by the dynamic client, so
+			// leave it empty. We want to list this resource in all
+			// namespaces if it's namespace scoped, so leave
+			// APIResource.Namespaced as false is all right.
+			apiResource := metav1.APIResource{Name: resource.Resource}
+			return client.ParameterCodec(dynamic.VersionedParameterEncoderWithV1Fallback).
+				Resource(&apiResource, v1.NamespaceAll).
+				Watch(&options)
+		},
+	}
+}
+
+func (gc *GarbageCollector) monitorFor(resource schema.GroupVersionResource, kind schema.GroupVersionKind) (monitor, error) {
+	// TODO: consider store in one storage.
+	glog.V(6).Infof("create storage for resource %s", resource)
+	var monitor monitor
+	client, err := gc.metaOnlyClientPool.ClientForGroupVersionKind(kind)
+	if err != nil {
+		return monitor, err
+	}
+	gc.registeredRateLimiterForMonitors.registerIfNotPresent(resource.GroupVersion(), client, "garbage_collector_monitoring")
+	setObjectTypeMeta := func(obj interface{}) {
+		runtimeObject, ok := obj.(runtime.Object)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("expected runtime.Object, got %#v", obj))
+		}
+		runtimeObject.GetObjectKind().SetGroupVersionKind(kind)
+	}
+	monitor.store, monitor.controller = cache.NewInformer(
+		gcListWatcher(client, resource),
+		nil,
+		ResourceResyncTime,
+		cache.ResourceEventHandlerFuncs{
+			// add the event to the propagator's eventQueue.
+			AddFunc: func(obj interface{}) {
+				setObjectTypeMeta(obj)
+				event := &event{
+					eventType: addEvent,
+					obj:       obj,
+				}
+				gc.propagator.eventQueue.Add(&workqueue.TimedWorkQueueItem{StartTime: gc.clock.Now(), Object: event})
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				setObjectTypeMeta(newObj)
+				event := &event{updateEvent, newObj, oldObj}
+				gc.propagator.eventQueue.Add(&workqueue.TimedWorkQueueItem{StartTime: gc.clock.Now(), Object: event})
+			},
+			DeleteFunc: func(obj interface{}) {
+				// delta fifo may wrap the object in a cache.DeletedFinalStateUnknown, unwrap it
+				if deletedFinalStateUnknown, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+					obj = deletedFinalStateUnknown.Obj
+				}
+				setObjectTypeMeta(obj)
+				event := &event{
+					eventType: deleteEvent,
+					obj:       obj,
+				}
+				gc.propagator.eventQueue.Add(&workqueue.TimedWorkQueueItem{StartTime: gc.clock.Now(), Object: event})
+			},
+		},
+	)
+	return monitor, nil
+}
+
+var ignoredResources = map[schema.GroupVersionResource]struct{}{
```

Comment:
I know this is pre-existing, but make a note to pull this information from the discovery doc.